File: base/src/main/java/de/hipp/pnp/converter/BaseConverter.java
package de.hipp.pnp.converter;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.apache.logging.log4j.util.Strings;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.persistence.AttributeConverter;
import java.io.IOException;

abstract class BaseConverter<Y> implements AttributeConverter<Y, String> {

    ObjectMapper objectMapper = new ObjectMapper();
    Logger logger = LoggerFactory.getLogger(this.getClass());
    Y type;

    @Override
    public String convertToDatabaseColumn(Y customerInfo) {

        String customerInfoJson = null;
        try {
            customerInfoJson = objectMapper.writeValueAsString(customerInfo);
        } catch (final JsonProcessingException e) {
            logger.error("JSON writing error", e);
        }

        return customerInfoJson;
    }

    @Override
    public Y convertToEntityAttribute(String customerInfoJSON) {

        Y valueInfo = null;
        if (Strings.isBlank(customerInfoJSON)) {
            return null;
        }
        try {
            valueInfo = objectMapper.readValue(
                    customerInfoJSON,
                    objectMapper.getTypeFactory().constructType(type.getClass())
            );
        } catch (final IOException e) {
            logger.error("JSON reading error", e);
        }

        return valueInfo;
    }
}

File: src/main/kotlin/de/hipp/pnp/base/converter/persistent/BaseConverter.kt
package de.hipp.pnp.base.converter.persistent

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import jakarta.persistence.AttributeConverter
import org.apache.logging.log4j.util.Strings
import io.github.oshai.kotlinlogging.KotlinLogging
import java.io.IOException

abstract class BaseConverter<Y> : AttributeConverter<Y, String> {

    private val objectMapper = ObjectMapper()
    protected val logger = KotlinLogging.logger {}

    protected var type: Y? = null

    override fun convertToDatabaseColumn(customerInfo: Y?): String? {
        var customerInfoJson: String? = null
        try {
            customerInfoJson = objectMapper.writeValueAsString(customerInfo)
        } catch (e: JsonProcessingException) {
            logger.error(e) { "JSON writing error" }
        }
        return customerInfoJson
    }

    override fun convertToEntityAttribute(customerInfoJSON: String?): Y? {
        var valueInfo: Y? = null
        if (Strings.isBlank(customerInfoJSON)) {
            return null
        }
        try {
            type?.let {
                valueInfo = objectMapper.readValue(
                    customerInfoJSON,
                    objectMapper.typeFactory.constructType(it::class.java)
                )
            }
        } catch (e: IOException) {
            logger.error(e) { "JSON reading error" }
        }
        return valueInfo
    }

    fun setType(type: Y): BaseConverter<Y> {
        this.type = type
        return this
    }
}